torch/_custom_op.py

```diff
@@ -0,0 +1,346 @@
+import inspect
+import typing
+import weakref
+
+from torchgen.model import FunctionSchema, OperatorName, SchemaKind
+
+import torch
+import torch._C as _C
+import torch.library as library
+import torch.utils._pytree as pytree
+
+"""
+There are various APIs for defining custom-operator-like things in PyTorch:
+- [user-facing] autograd.Function (Python)
+- [user-facing] custom_op (Python)
+- [for power users] torch.library (Python)
+- [for power users] TORCH_LIBRARY (C++)
+
+This file contains the implementation for a Simple Custom Operator API (CustomOp).
+Using CustomOp, you are able to define a custom operator and implement interactions
+between the CustomOp and various PyTorch subsystems, including all the subsystems
+that are necessary for a custom operator to work with torch.compile (i.e.,
+autograd, meta, functionalization).
+
+CustomOp is positioned as being safer and easier to use than
+torch.library/TORCH_LIBRARY, which require deep understanding of PyTorch internals.
+In additional, it supports torch.compile better than and is in general more
+comprehensive than autograd.Function, which only supports implementing gradient
+computation and vmap rules.
+"""
+
+__all__ = ["custom_op", "CustomOp"]
+
+
+SUPPORTED_DEVICE_TYPES = ("cpu", "cuda")
+
+DEVICE_TYPE_TO_KEY = {
+    "cpu": "CPU",
+    "cuda": "CUDA",
+}
+
+# We will not let users register CustomOps with anything that could look like
+# PyTorch internals to avoid confusion.
+RESERVED_NS = {
+    "prim",
+    "prims",
+    "aten",
+    "at",
+    "torch",
+    "pytorch",
+}
+
+
+def custom_op(schema: str, *, ns: str) -> typing.Callable:
+    r"""Creates a new CustomOp object.
+
+    In PyTorch, defining an op (short for "operator") is a two step-process:
+    - we need to define (create) the op
+    - we need to implement behavior for how the operator interacts with
+      various PyTorch subsystems, like CPU/CUDA Tensors, Autograd, etc.
+
+    This entrypoint defines the CustomOp object (the first step);
+    you must then perform the second step by calling various methods on
+    the CustomOp object.
+
+    This API is used as a decorator (see examples).
+
+    Arguments:
+        schema (str): The schema of the CustomOp.
+        ns (str): The namespace of the CustomOp. PyTorch operators need a
+            namespace; a given operator may only be created once. If you
+            are writing a Python library, we recommend the namespace to be
+            the name of your top-level module.
+
+    Example::
+        >>> import numpy as np
+        >>>
+        >>> # Step 1: define the CustomOp.
+        >>> # We need to provide the decorator a "prototype function"
+        >>> # (a function with Python ellipses as the body).
+        >>> @custom_op('(Tensor x) -> Tensor')
+        >>> def numpy_sin(x):
```

Comment:
Unfortunately, no :P. Prototype function is just there so we can get the name/module

I can add some check (that the number of arguments and the name of the arguments are the same?) if you think it would be nicer

Reply:
Oh wait this is the API you proposed in your doc initially lol, why did we switch to this new version?

Reply:
@ezyang, some feedback that we should go back to the original API lol.

@soulitzer, the reason why we switched to the new version is the `__module__`/`__name__` aren't set "correctly". That is, if you do the following, then `numpy_sin.__name__` is None and `numpy_sin.__module__` is torch._custom_op.

```py
numpy_sin = custom_op('numpy_sin(Tensor x) -> Tensor')
```

Now, the problem is mostly that the user may expect `numpy_sin` to behave like a function (it will be interposable by torch_function), but it doesn't really look like a function (it doesn't have `__name__` or `__module__` set like a function). 

At this point it's a tradeoff discussion. We can confuse the user one way or the other, so what is less confusing?

Reply:
A secondary reason why we switched, but I don't think will actually happen in the future, is that we may be able to infer a schema from the prototype function's signature. But I wouldn't count on this, because there's not an existing way to capture alias information in the schema, and if we ever need to express more information in the schema we may have a problem:
```
@custom_op(ns='_torch_testing')
def numpy_sin(x: Tensor) -> Tensor
    ...
```

Reply:
Ah thanks for the context.

Re: inferring the schema, since view or inplace is not common, could we make specifying the schema optional unless we detect that the user tries to do either of those things? That would make it possible to infer the schema for most cases right?

Reply:
> IMO it is better if we drill into the user that it is possible for their schema to be wrong (and have them read through a doc page on how to write a schema and what it consists of), instead of us detecting it ourselves, because it's a bit difficult to detect ourselves:

Yeah I agree with this

Given that the secondary reason is unlikely to materialize. Would it resolve the issue if we just set __name__ and __module__ ourselves?

Reply:
> Would it resolve the issue if we just set __name__ and __module__ ourselves?

Well, we'd need to ask our users to set __name__ and __module__, right? Inside of the `custom_op` function, when we construct the CustomOp object, we do not know the __module__ of `custom_op`'s caller

Reply:
@soulitzer I am only going to check that the number of arguments and the name of the arguments are correct. Before we make this API public, we will decide if we want to autogenerate the schema from the function signature or not. In the latter case, we may need additional checks: that the type annotations in the function signature are compatible with the schema types.

Reply:
I don't know if we want to go there, but note that we could set the `__name__` by looking at the schema, and we could set the module by looking at `inspect.getmodule(inspect.stack()[1][0])`. Now, it's reasonable if you guys think that this is a bit too much.